java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,887 @@
+    }
+    return null;
+  }
+
+  public void get(int index, DenseUnionHolder holder) {
+    FieldReader reader = new DenseUnionReader(DenseUnionVector.this);
+    int offset = offsetBuffer.getInt(index * OFFSET_WIDTH);
+    reader.setPosition(offset);
+    holder.reader = reader;
+  }
+
+  public int getValueCount() {
+    return valueCount;
+  }
+
+  public boolean isNull(int index) {
+    return (typeBuffer.getByte(index * TYPE_WIDTH) == 0);
+  }
+
+  @Override
+  public int getNullCount() {
+    int nullCount = 0;
+    for (int i = 0; i < getValueCount(); i++) {
+      if (isNull(i)) {
+        nullCount++;
+      }
+    }
+    return nullCount;
+  }
+
+  public int isSet(int index) {
+    return isNull(index) ? 0 : 1;
+  }
+
+  DenseUnionWriter writer;
+
+  public void setValueCount(int valueCount) {
+    this.valueCount = valueCount;
+    while (valueCount > getTypeBufferValueCapacity()) {
+      reallocTypeBuffer();
+      reallocOffsetBuffer();
+    }
+    internalStruct.setValueCount(valueCount);
+  }
+
+  public void setSafe(int index, DenseUnionHolder holder) {
+    FieldReader reader = holder.reader;
+    if (writer == null) {
+      writer = new DenseUnionWriter(DenseUnionVector.this);
+    }
+    int offset = offsetBuffer.getInt(index * OFFSET_WIDTH);
+    writer.setPosition(offset);
+    MinorType type = reader.getMinorType();
+    switch (type) {
+      <#list vv.types as type>
+        <#list type.minor as minor>
+          <#assign name = minor.class?cap_first />
+          <#assign fields = minor.fields!type.fields />
+          <#assign uncappedName = name?uncap_first/>
+          <#if !minor.typeParams?? >
+      case ${name?upper_case}:
+      Nullable${name}Holder ${uncappedName}Holder = new Nullable${name}Holder();
+      reader.read(${uncappedName}Holder);
+      setSafe(index, ${uncappedName}Holder);
+      break;
+          </#if>
+        </#list>
+      </#list>
+      case STRUCT: {
+        ComplexCopier.copy(reader, writer);
+        break;
+      }
+      case LIST: {
+        ComplexCopier.copy(reader, writer);
+        break;
+      }
+      default:
+        throw new UnsupportedOperationException();
+    }
+  }
+    <#list vv.types as type>
+      <#list type.minor as minor>
+        <#assign name = minor.class?cap_first />
+        <#assign fields = minor.fields!type.fields />
+        <#assign uncappedName = name?uncap_first/>
+        <#if !minor.typeParams?? >
+  public void setSafe(int index, Nullable${name}Holder holder) {
+    setType(index, MinorType.${name?upper_case});
+    while (index >= getOffsetBufferValueCapacity()) {
+      reallocOffsetBuffer();
+    }
+    ${name}Vector vector = get${name}Vector();
+    int offset = vector.getValueCount();
+    vector.setValueCount(offset + 1);
+    vector.setSafe(offset, holder);
+    offsetBuffer.setInt(index * OFFSET_WIDTH, offset);
+  }
+
+        </#if>
+      </#list>
+    </#list>
+
+  public void setType(int index, int typeId) {
```

Comment:
typeIds should probably be "byte" everywhere.

Reply:
Revised accordingly. Thanks.